samples/nrf9160/mqtt_simple/src/main.c

```diff
@@ -0,0 +1,284 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <uart.h>
+#include <string.h>
+#include <at_host.h>
+
+#include <net/mqtt.h>
+#include <net/socket.h>
+#include <lte_lc.h>
+
+#define APP_MQTT_BUFFER_SIZE 128
+
+/* Buffers for MQTT client. */
+static u8_t rx_buffer[APP_MQTT_BUFFER_SIZE];
+static u8_t tx_buffer[APP_MQTT_BUFFER_SIZE];
+
+/* The mqtt client struct */
+static struct mqtt_client client;
+
+/* MQTT Broker details. */
+static struct sockaddr_storage broker;
+
+static bool connected;
+
+static void data_print(u8_t *prefix, u8_t *data, size_t len)
+{
+	char buf[len + 1];
+
+	memcpy(buf, data, len);
+	buf[len] = 0;
+	printk("%s%s\n", prefix, buf);
+}
+
+static int data_publish(struct mqtt_client *c, enum mqtt_qos qos,
+	u8_t *data, size_t len)
+{
+	struct mqtt_publish_param param;
+
+	param.message.topic.qos = qos;
+	param.message.topic.topic.utf8 = CONFIG_MQTT_PUB_TOPIC;
+	param.message.topic.topic.size = strlen(CONFIG_MQTT_PUB_TOPIC);
+	param.message.payload.data = data;
+	param.message.payload.len = len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+	param.retain_flag = 0;
+
+	data_print("Publish: ", data, len);
+	printk("to topic: %s len: %d\n",
+		CONFIG_MQTT_PUB_TOPIC,
+		strlen(CONFIG_MQTT_PUB_TOPIC));
+
+	return mqtt_publish(c, &param);
+}
+
+static int subscribe(void)
+{
+	struct mqtt_topic subscribe_topic = {
+		.topic = {
+			.utf8 = CONFIG_MQTT_SUB_TOPIC,
+			.size = strlen(CONFIG_MQTT_SUB_TOPIC)
+		},
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE
+	};
+
+	const struct mqtt_subscription_list subscription_list = {
+		.list = &subscribe_topic,
+		.list_count = 1,
+		.message_id = 1234
+	};
+
+	printk("Subscribing to: %s len %d\n", CONFIG_MQTT_SUB_TOPIC,
+		strlen(CONFIG_MQTT_SUB_TOPIC));
+
+	return mqtt_subscribe(&client, &subscription_list);
+}
+
+void mqtt_evt_handler(struct mqtt_client *const c,
+		      const struct mqtt_evt *evt)
+{
+	switch (evt->type) {
+	case MQTT_EVT_CONNACK:
+		if (evt->result != 0) {
+			printk("MQTT connect failed %d\n", evt->result);
+			break;
+		}
+
+		connected = true;
+		printk("[%s:%d] MQTT client connected!\n", __func__, __LINE__);
+		subscribe();
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		printk("[%s:%d] MQTT client disconnected %d\n", __func__,
+		       __LINE__, evt->result);
+
+		connected = false;
+		break;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *p = &evt->param.publish;
+		u8_t buf[p->message.payload.len];
+		int len;
+
+		printk("[%s:%d] MQTT PUBLISH result=%d len=%d\n", __func__,
+		       __LINE__, evt->result, p->message.payload.len);
+		len = mqtt_read_publish_payload(c, buf, sizeof(buf));
```

Comment:
This won't work in general case, this function has similar API to a non-blocking socket `recv` function, it's a reads a stream hence it might not read the entire payload at once. You should call `mqtt_read_publish_payload` in a loop until you read entire payload (`p->message.payload.len`), preferably with `poll` inbeetween, see https://github.com/NordicPlayground/fw-nrfconnect-nrf/blob/5be3bb964eaf9f2269537972a3dc9b1d8b3f73c1/subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c#L444

Reply:
will fix